src/compiler/cpp_generator.cc

```diff
@@ -547,6 +548,115 @@ void PrintHeaderClientMethod(grpc_generator::Printer* printer,
   }
 }
 
+void PrintHeaderClientMethodCallbackInterfacesStart(
+    grpc_generator::Printer* printer,
+    std::map<grpc::string, grpc::string>* vars) {
+  // This declares the interface for the callback-based API. The components
+  // are pure; even though this is new (post-1.0) API, it can be pure because
+  // it is an entirely new interface that happens to be scoped within
+  // StubInterface, not new additions to StubInterface itself
+  printer->Print("class experimental_async_interface {\n");
+  // All methods in this new interface are public. There is no need for private
+  // "Raw" methods since the callback-based API returns unowned raw pointers
+  printer->Print(" public:\n");
+  printer->Indent();
+}
+
+void PrintHeaderClientMethodCallbackInterfaces(
+    grpc_generator::Printer* printer, const grpc_generator::Method* method,
+    std::map<grpc::string, grpc::string>* vars, bool is_public) {
+  // Reserve is_public for future expansion
+  assert(is_public);
+
+  (*vars)["Method"] = method->name();
+  (*vars)["Request"] = method->input_type_name();
+  (*vars)["Response"] = method->output_type_name();
+
+  if (method->NoStreaming()) {
+    printer->Print(*vars,
```

Comment:
Consider adding an overload for preserialized requests (via ::grpc::ByteBuffer).

Reply:
So the overload method works for unary, but not for streaming (esp bidi streaming). Agreed that this could be interesting. @ncteisen mentioned this as part of a broader proposal, but nobody ever gave a use case for this part. Would be interesting to see a use-case. That said, that's also independent of callback vs non-calback.

Reply:
Actually, the streaming case isn't necessarily bad either since we can just have overloads of Read and Write, I think.

Reply:
That said, I think "Let's push this to a later time/PR"